branchandbound.js

```diff
@@ -0,0 +1,143 @@
+var utils = require('./utils')
+
+var maxTries = 1000000
+
+function calculateEffectiveValues (utxos, feeRate) {
+  return utxos.map(function (utxo) {
+    if (isNaN(utils.uintOrNaN(utxo.value))) {
+      return {
+        utxo: utxo,
+        effectiveValue: 0
+      }
+    }
+
+    var effectiveFee = utils.inputBytes(utxo) * feeRate
+    var effectiveValue = utxo.value - effectiveFee
+    return {
+      utxo: utxo,
+      effectiveValue: effectiveValue
+    }
+  })
+}
+
+module.exports = function branchAndBound (utxos, outputs, feeRate) {
+  if (!isFinite(utils.uintOrNaN(feeRate))) return {}
+
+  var costPerOutput = utils.outputBytes({}) * feeRate
+  var costPerInput = utils.inputBytes({}) * feeRate
+  var costOfChange = costPerInput + costPerOutput
+
+  var outAccum = utils.sumOrNaN(outputs)
+
+  var effectiveUtxos = calculateEffectiveValues(utxos, feeRate).filter(function (x) {
+    return x.effectiveValue > 0
+  }).sort(function (a, b) {
+    return b.effectiveValue - a.effectiveValue
+  })
+
+  var selected = search(effectiveUtxos, outAccum, costOfChange)
+  if (selected != null) {
+    var inputs = []
+
+    for (var i = 0; i < effectiveUtxos.length; i++) {
+      if (selected[i]) {
+        inputs.push(effectiveUtxos[i].utxo)
+      }
+    }
+
+    return utils.finalize(inputs, outputs, feeRate)
+  } else {
+    const fee = feeRate * utxos.reduce(function (a, x) {
```

Comment:
It could be that now that the transaction size is properly estimated, the allowance window for the exact match actually is too big. You might want to try with a "costOfChange" that is actually only the saved change output's cost. That would be the lower bound on the saved cost, but may cause the algorithm to find fewer matches or require more inputs to find matches.